### DISCUSSION TITLE
Enable Response Buffering by default in Umbraco v9

### DISCUSSION THREAD

**Discussion by mattbrailsford** [umbraco/Umbraco-CMS#11122] [Features and ideas]
This is really going to polarize people (I know it already has in the Discord channel) but I want to see if we can discuss enabling response buffering by default in Umbraco v9.

**Background**

I've recently been setting up Vendr in Umbraco v9, building out a site and getting things working when I hit upon a problem. Something that worked in v8 didn't work in v9 anymore and this ultimately boils down to calling a method in my view that performs a task and then attempts to write a cookie. In v8 (NF) this all works fine, but in v9 (NC) this threw an exception.

    Headers are read-only, response has already started

After some investigation this appears to stem from the fact that in .NET Core the response is now returned in chunks, with the headers thus being the first thin that gets sent down the line, meaning that when the view starts to render, the headers have already been sent and so can no longer be modified and so the code from the view errors. This differs to v8 (NF) where responses would be buffered by default and so nothing would be sent until the whole page is rendered and thus anything could be changed right up to the request being sent.

In my investigations I found there were 2 ways to solve this issue:

1. Move all logic calls to a Controller / Component such that only static variables are sent to the view.
2. Install the package `Microsoft.AspNetCore.Buffering` and add `app.UseResponseBuffering();` to the `Startup.Configure` method to re-anable response buffering.

After a lengthy discussion on Discord, even though option 1 is the better practice,  I'm now wondering if we should go with option 2 and enable response buffering by default?

**Chunking vs Buffering**
Currently the default in v9 / .NET Core is for responses to be chunked and so a response is sent to the browser as quickly as possible. This has the advantage of creating a great Time to First Byte (TTFB) response and allows the browser to start loading client resources as quickly as possible. With buffering though, nothing reaches the browser until the whole page is sent down the line and so it can take pages that little bit longer to start loading those resources.

There is also the argument that streaming takes up less server resources than buffering which needs to load the whole page into memory before sending it down the line.

**Why enable buffering?**
Considering the above, why would we even consider enabling buffering by default if chunking is more optimal? Well, my answer to this is Developer Experience (DX), and specifically beginners and "implementors".

With chunking, whilst it offers the best performance, it comes with some side effects. Because some resources get sent before views start to render, it now starts to pose problems for those that write logic into views. Yea, I know, we shouldn't be writing logic in views, but some times you just need to do something fast, or you just want to learn how something works, and other times you just need something done now. Umbraco has long had a history of supporting these devs and I think it is worth continuing with this.

But should we really introduce a perf issue for the sake of some devs following bad practice? Well, the question really is "how significant is the perf difference?" and from some historic issue on this, it appears not massively https://github.com/dotnet/aspnetcore/issues/3100

This is a pretty old issue though so I don't know how this stacks up today, but it really shouldn't be getting any worse than this.

In addition to this, there is already something logged for .NET 7 to improve this in some way (either bring back buffering, or better handling, it's not clear) https://github.com/dotnet/aspnetcore/issues/27470

**Proposition**
My proposition then is that given there may be improvements coming, and it's clear there are side effects to chunking that require more developer centric knowledge to deal with (ie, moving code to controllers / components), could we enable response buffering by default such that people just starting with Umbraco get a simple and friendly dev experience but then expose a means of disabling buffering once they are at the best practices stage or if it's being worked on by a dev team that know how to manage the new approach and already follow those best practices.

There is an argument that we should maybe make it configurable to "enable buffering", but I personally think this would be a mistake as you are then requiring beginners to do this, before they can have a nice experience. I really think it would be much better for them to get the simple experience, and those in the know wanting to get maximum performance can then disable it and handle the side effects.

Well, there it is, now how do I mute this thread? 😂

- Thanks to Dave de Moel for finding the perf / .NET 7 issues 👍

**Comment by protherj** [comment]
OK, I'm coming around to have this enabled by default.  That would make a better experience to those new to Umbraco or looking for a more flexible experience.

I wonder if Umbraco could have a configuration of "mode" that would enable the things behind the scenes (like buffering or whatever else we find) that allows devs to use Umbraco like Matt indicates.  Like a "strict" or "flexible" setting.   By default it would be set to "flexible" which would enable buffering.

Just an idea because it may be more friendly for devs to find and be aware of a mode or configuration than the "buffering" .NET configuration setting which is a bit hidden IMO and it could include more of these type of "gotcha" settings in the future.

Thanks for listening :)

**Reply by hfloyd** [reply]
If those of you familiar with .Net Core (and its ongoing evolution) can foresee additional "advanced" settings coming up, then having a single Umbraco-managed place for them all would be helpful. However, it would be better for the actual individual settings be separately configured, as much as is possible. For instance, some other advanced feature may become available which a Dev wants to turn "on" - but they don't want to turn on "chunking"...

One of the "hidden" benefits of using Umbraco is that really knowledgeable and smart devs who understand these sorts of advanced coding things are thinking deeply about best practices in the broadest way (pros and cons, performance versus flexibility, etc.) and building that into the CMS, so that other devs/website implementors can focus on the unique business and technical requirements of their websites. For instance, as an Umbraco Developer, I don't have to "worry" much about caching - because it is built-in. This is a major benefit to Umbraco.

**Reply by bergmania** [reply]
Personally, I like the idea of strict and flexible modes, but I fear using flexible by default would lead to the Umbraco Codebase or packages to rely on the features in "flexible" and thereby limit professional work with Umbraco, that would require the strict mode.

IMO the power of Umbraco is that is can be used for both enterprise solutions and also hobby projects.  I fear a default flexible mode would lead to Umbraco can only be used for hobby projects. That said, I think we should do as much as possible to make it easy for hobby projects, as long as we do not limit the use for professional work.

**Reply by mattbrailsford** [reply]
Whilst I’m not averse to the flexible/strict mode settings, I’m concerned it’s stretching the scope of this issue and creating concerns over what may possibly be. 

If the mode is wide sweeping I could see things like pkg devs targeting a particular mode an issue, but if we just take the original concern on its own I don’t think it’s as big a worry. 

In this particular case, I can’t see how it can be targeted by pkg devs as a requirement? I really can’t see a way in Vendr that we could allow method calls in a views that modify the users session in any other way, so it’s not through lack of wanting. In this particular case, the only real workaround is to force the implementors to use best practice and move all logic to controllers/components (or install the extra pkg and configure response buffering) which puts the burden on them. 

I think if we maintain focus on that, and what it means to the average dev, it’s a little less concerning. 

I also want to be clear, code in views != hobbyists. I have created many a “professional” site in this way in the past as I know do others, so I think it would be a mistake to think it’s only the minority of people that don’t know better yet that would be affected.

**Reply by bergmania** [reply]
I agree, "hobby project" is not perfect wording, but it does change the idea. 

If buffering is turned on by default I fear packages rely on this to be turned on. If a Umbraco developer tries to optimize for performance they would like to turn off buffering, but if some of the packages they use rely on it, they are forced to keep using buffering.

**Reply by mattbrailsford** [reply]
But I'm not quite clear how a package dev is to rely on it? What are some concrete examples? Otherwise we just have concerns over something that we have no idea would/could happen?

In my particular example where it affects writing cookies, the implementor isn't forced to keep response buffering on, only if they want to disable it, they must move their logic to controllers / components. Whilst yes, it puts extra burden on them then to refactor at that point (if they aren't already using best practice), but really, this is what we are asking ALL devs to do now, regardless of needs and ability if we continue with the default chunking approach.

**Reply by bergmania** [reply]
The smallest example I can come up with is like this.

If your package uses middleware. Eg. this simple example

```c#
app.Use(async (context, next) =>
{
    await next();

    context.Response.Headers["X-TooLate"] = "This is header is set too late by default";
});
```

This middleware will fail by default, because the response headers are added after `await next();`
But if `app.UseResponseBuffering()` is added before this middleware, the middleware works perfectly well and sets the header.

So packages could implement there middleware in a way that only works if `app.UseResponseBuffering()` is added before, and thereby forcing users of their package to use response buffering.

**Reply by mattbrailsford** [reply]
Fair example 👍

I'm still wondering though who is the best person to put this responsibility on though? Implementors who have varied skill levels, wondering why a package isn't working because they chose to code in views? or package devs, who arguably should be ok learning this (via docs) and ensuring their packages can be used in either way, or if it can't directly, can document the side-effects?

**Reply by marcemarc** [reply]
> IMO the power of Umbraco is that it can be used for both enterprise solutions and also hobby projects.

In the real world, there is really 'quite a lot' in-between 'enterprise' and 'hobby', it's brave to only see it in those black and white terms.

**Reply by bergmania** [reply]
> > IMO the power of Umbraco is that it can be used for both enterprise solutions and also hobby projects.
> 
> In the real world, there is really 'quite a lot' in-between 'enterprise' and 'hobby', it's brave to only see it in those black and white terms.

I of course also mean everything in between

**Reply by hfloyd** [reply]
From the perspective of someone who spends all her time in that broad place "in-between"... 

I work with many different types of developers and teams at various sizes of organizations (both "end-user in-house" as well as "agency"), so I have got a pretty good sense of the various levels of skills of people who are working on Umbraco sites. Additionally, most of the projects I work on I didn't get to design from the beginning - I am being brought on because the existing site isn't working well, or the original developers are no longer available, so I see how others are building sites. 

Out of the dozens of site I have looked-at/worked on, I think only one had separate Controllers defined for each View (incidentally, this is also the most problematic site I have worked on - buggy, overly complicated, non-standard Umbraco practices, and difficult to change). 

For the majority of website projects, there is a big risk associated with "over-engineering" (which adds costs, complexity, and inflexibility) - which is why I think so few Umbraco websites I have encountered tend to use a "pure" MVC setup.

Personally, when I am involved in a project, I create a separate ".Core" class project next to the Web project, and I use ModelsBuilder + a site-wide "Base" interface to put a fair amount of general logic into the models that come into a View... but I rarely have "per-view" separate Controllers. 

Partly this is because I can't always count on my developer colleagues on any given project to be comfortable with editing / re-compiling a class project just to slightly change the filtering or sort order of the items displayed in a list on a page, etc. And partly because I don't want to over-complicate things as development moves along swiftly - If I add a new Doctype, a quick ModelsBuilder re-generation/re-build and I can work with the new data in a view immediately.

So, I strongly stand behind the idea that Umbraco Developers should NOT be forced to create 1-1 Controllers-Views. That being said, usually I don't need to manipulate things "outside" the visible rendered page (cookies, session, headers), so if this issue of "buffering vs chunking" only affects those items, then I am more open to the possibility of making the chunking the default option - with some nice documentation to describe the issue and how one would go about setting a cookie properly, for instance. If the issues affects basic things one might do in a View, then I would be against it.

BTW - From the preceding paragraph it should be obvious my level of ignorance regarding this specific issue of buffering in .Net Core. Keep in mind that probably 80%+ of other Umbraco Developers will have a similar level of ignorance, and plan accordingly. 😉

**Reply by protherj** [reply]
I love the discussion on this.  I agree that if this cookie / buffering issue is the only instance like this that could be hidden frustration, then having a strict/flexible mode doesn't make a whole lot of sense.  For now it seems like documentation and perhaps addressing it in training is the best idea.

Package devs should probably be developing defensively and always assume buffering is off so it works both ways for implementers.  

If we find lots of edge cases like this during the process of upgrade packages and sites, then perhaps revisiting a simple "mode" would make sense.  That would reduce the burden of documenting all of them effectively (who reads the docs anyways, right? ;) ).  But I agree with Matt that it is likely premature for that kind of feature for just a single issue.

**Reply by JasonElkin** [reply]
I think that's an important point to bear in mind. We're not talking about the virtues of code in views vs. viewmodels/controllers, we're talking about the fact that some methods that could be called in views in v8 can't in v9.

It'd be helpful to understand the scope of what methods are not longer available and whether that warrants such a significant change. 

We've always been in a situation where there are some methods/APIs that we just can't use in views. Sometimes this is an Umbraco decision, sometimes a framework decision.

What I'm worried about is reversing a framework decision that affects every single HTTP request in a site for the sake of a few methods that will likely be used very sparsely in any given site.

Selectively enabling this feature on a per-request (per template?) basis seems a much more sensible way to enable the now broken APIs than changing how HTTP responses work everywhere.

**Reply by kjac** [reply]
I'm sorry if this comes off as elitist or unfairly favoring experienced devs... but I truly believe that implementors bear the responsibility of knowing and understanding their tech stack; in this case be knowledgeable enough of .net (core or not) to construct a controller based solution to something like this.

In my opinion we should never compromise on any critical parts of Umbraco (i.e. performance, security, ...) to cater for bad practices. Instead we should be concerned with helping devs through documentation of common pitfalls (if this is indeed a common pitfall).

**Comment by JasonElkin** [comment]
As we're talking about the API the browser uses to fetch the website I think we need to be super careful to make sure we understand the real-world performance implications before a decision is made.

i.e. testing with sites with realistic frontend payloads and server-side code, across HTTP/1.1, HTTP/2 and... 🥁 HTTP/3! I suspect if there are performance considerations HTTP3 is where they would show up as that actually multiplexes the frames and sends them asynchronously.
